feat(wip): add YouTube OAuth

kaf-lamed-beyt opened the pull request:
@zcesur, I've been able to setup the OAuth flow with Ueberauth, but I'm having a blocker when trying to persist the state, via the callback.

Kindly take a look and let me know where I might be doing something wrong.

Settings UI on Algora

![Screenshot from 2024-09-24 11-22-12](https://github.com/user-attachments/assets/672290a0-6d29-4003-9d00-fe4dee8f1a57)

OAuth screen
![google-oauth](https://github.com/user-attachments/assets/03908e74-e1fe-4e0e-a8ea-150623f34523)

/claim #104

zcesur commented:
Looks good!

> I'm having a blocker when trying to persist the state, via the callback

Are there any errors in your logs when the callback is invoked?

kaf-lamed-beyt commented:
Yeah, there were. Something around the function `create_or_update_youtube_identity` not being defined. Hole on. I'll share a screenshot.

kaf-lamed-beyt commented:
Here @zcesur 

![image](https://github.com/user-attachments/assets/81a09768-2da8-4684-a55d-7dfe6bfad216)

zcesur commented:
We need to define the changeset we're using in the callback, see `lib/algora/accounts/identity.ex`

kaf-lamed-beyt commented:
Oh okay! I'll do that now.

kaf-lamed-beyt commented:
Had a couple of issues with retrieving the provider value. Apparently, Ueberauth returns the value as an atom. I had to convert it so it matches the `"identities"` schema. 

Clicking on the button when an account is connected automatically disconnects that identity. Is this a good thing to do? 

I've got a slight issue with the placement of the button though. Although, I added a margin, it did not align the button correctly. PS: the other button doesn't show up when it is connected. 
![Screenshot from 2024-09-25 15-05-02](https://github.com/user-attachments/assets/1458b3f5-58fd-4145-8b18-e63fdcf853a4)

zcesur commented on lib/algora/accounts/user.ex:
```diff
@@ -49,6 +50,37 @@ defmodule Algora.Accounts.User do
+      provider_email: auth.info.email,
+      provider_login: auth.info.email,
+      provider_token: auth.credentials.token,
+      provider_refresh_token: auth.credentials.refresh_token,
```
I think Google doesn't provide a refresh token unless we request one:

```elixir
credentials: %Ueberauth.Auth.Credentials{
  token: "...",
  refresh_token: nil,
  token_type: "Bearer",
  expires_at: 1727452355,
}
```

Given that the access token expires in 1 hour, we'll definitely need a refresh token

kaf-lamed-beyt replied:
oh okay. I'll look at this too.

zcesur commented on lib/algora/accounts/user.ex:
```diff
@@ -49,6 +50,37 @@ defmodule Algora.Accounts.User do
   end
 
+  def create_or_update_youtube_identity(user, auth) do
+    attrs = %{
```
Let's also persist `auth.info` inside `provider_meta`

zcesur reviewed (commented):
Almost there!

> Clicking on the button when an account is connected automatically disconnects that identity. 

This might be confusing so let's remove that logic

kaf-lamed-beyt commented:
> This might be confusing so let's remove that logic

Alright, I'll have this removed.

kaf-lamed-beyt commented on config/config.exs:
```diff
@@ -72,6 +72,17 @@ config :phoenix, :json_library, Jason
+# ueberauth config
+config :ueberauth, Ueberauth,
+  providers: [
+    google: {Ueberauth.Strategy.Google, [default_scope: "email", scope: "youtube youtube.upload yt-analytics.readonly"]}
```
Huh! i think it's because I passed it wrong.

I'll move the others into `"default_scope"` and see how that goes.

kaf-lamed-beyt commented:
So, a little update here @zcesur. 

Yes. the scopes were not exactly correct. So I've updated it, and now the OAuth screen shows this properly.
![oauth-consent-screen](https://github.com/user-attachments/assets/f9b8c947-cfbb-4b2d-8d6b-9610f5cde1f8). 

The `refresh_token` property returns a value now &mdash; instead of nil. 

I added a dummy function to update the Google token... basically just followed your approach in `/lib/algora/accounts.ex`.But I realized that I might have to implement something similar to the one here for Restream 👇🏼 

https://github.com/algora-io/tv/blob/677d6c74b86d853200aea1dfc7624b4a8f74f823/lib/algora/restream.ex#L66-L92

Would that be something I need to get into? If yes, perhaps, a little guidance, thanks.

zcesur commented:
> The `refresh_token` property returns a value now — instead of nil.

Perfect

> But I realized that I might have to implement something similar to the one here for Restream

Yeah we need this for offline access (viewing analytics, uploading clips etc). It's pretty much the same as refreshing Restream tokens, [check out the docs here](https://developers.google.com/identity/protocols/oauth2/web-server#offline)

<img src="https://github.com/user-attachments/assets/2ad080aa-82da-4a6d-9e78-669d54f5b20a" width="500" />

kaf-lamed-beyt commented:
Awesome! I'll check the docs out.

kaf-lamed-beyt commented:
@zcesur, i added the `refresh_token` function now. When you can, please a look and let me know if I'm on the right path.

zcesur commented:
Thanks Caleb!

kaf-lamed-beyt commented:
> Thanks Caleb!

Ayy! Glad I got to work on this. 🫡
